File: src/main/java/dark/leech/text/plugin/js/loader/PageLoader.java
package dark.leech.text.plugin.js.loader;

import java.util.ArrayList;
import java.util.List;

import org.mozilla.javascript.NativeArray;
import org.mozilla.javascript.NativeObject;

import dark.leech.text.action.Log;
import dark.leech.text.enities.PluginEntity;
import dark.leech.text.plugin.js.sandbox.JsSandbox;

/**
 * JavaScript page discovery loader for vBook plugins using Rhino. Executes the 'page' script and
 * parses result into a list of page URLs. Used for discovering all chapter page URLs for a novel
 * from HTML metadata (e.g., total-page input field).
 */
public class PageLoader extends AbstractLoader<List<String>> {

    private PageLoader(PluginEntity plugin) {
        super(plugin);
    }

    public static PageLoader with(PluginEntity plugin) {
        return new PageLoader(plugin);
    }

    @Override
    protected String getScript() {
        return plugin.getPageGetter();
    }

    @Override
    protected LoaderType getLoaderType() {
        return LoaderType.PAGE;
    }

    @Override
    protected List<String> processResult(Object result, String url) {
        if (result == null || result == org.mozilla.javascript.Undefined.instance) {
            Log.add("[PageLoader] Script returned null or undefined");
            return new ArrayList<>();
        }

        // Unwrap NativeJavaObject if present (Rhino wraps Java objects)
        if (result instanceof org.mozilla.javascript.NativeJavaObject nativeJavaObj) {
            result = nativeJavaObj.unwrap();
        }

        // Check for Response wrapper (vBooks compatibility)
        if (!Response.isSuccess(result)) {
            String errorMsg = Response.getErrorMessage(result);
            Log.add("[PageLoader] Error response: " + errorMsg);
            return new ArrayList<>();
        }

        Object data = Response.getData(result);

        Log.add("[PageLoader] Processing result, type: " + result.getClass().getName());
        List<String> urlList = convertToStringList(data);
        Log.add("[PageLoader] Discovered " + urlList.size() + " page URLs");
        return urlList;
    }

    /**
     * Load page URLs from the given URL using the configured script. This is a one-shot discovery
     * operation that returns all available page URLs at once.
     *
     * @param url Target URL (novel detail page)
     * @return List of discovered page URLs, or empty list on error
     */
    public List<String> load(String url) {
        String script = getScript();
        if (plugin == null || script == null) {
            Log.add("[PageLoader] No script available");
            return new ArrayList<>();
        }

        JsSandbox sandbox = null;
        try {
            // Extract base URL
            String baseUrl = plugin.getSource();
            if (baseUrl == null || baseUrl.isEmpty()) {
                baseUrl = extractBaseUrl(url);
            }

            // Create secure sandbox for PAGE loader
            sandbox =
                    new JsSandbox.Builder()
                            .loaderType(getLoaderType())
                            .baseUrl(baseUrl)
                            .targetUrl(url)
                            .build();

            // Execute plugin script in sandbox
            if (!sandbox.execute(script, getScriptName())) {
                return new ArrayList<>();
            }

            // Call execute(url) function in sandbox (no page parameter)
            Object result = sandbox.callFunction("execute", url);

            if (result == null || result == org.mozilla.javascript.Undefined.instance) {
                return new ArrayList<>();
            }

            return processResult(result, url);

        } catch (Exception e) {
            String errorMsg = e.getMessage();
            if (errorMsg == null) {
                errorMsg = e.getClass().getName();
            }
            Log.add("[PageLoader] JavaScript execution error: " + errorMsg);
            return new ArrayList<>();
        } finally {
            if (sandbox != null) {
                sandbox.close();
            }
        }
    }

    /**
     * Convert object to List of String URLs. Handles NativeArray, NativeObject, NativeJavaObject,
     * and raw arrays.
     */
    private List<String> convertToStringList(Object data) {
        if (data == null) {
            return new ArrayList<>();
        }

        if (data instanceof NativeArray) {
            NativeArray nativeArray = (NativeArray) data;
            long size = nativeArray.size();
            if (size > Integer.MAX_VALUE) {
                Log.add(
                        "[PageLoader] Array size exceeds Integer.MAX_VALUE, truncating to "
                                + Integer.MAX_VALUE);
                size = Integer.MAX_VALUE;
            }
            List<String> list = new ArrayList<>((int) size);
            for (int i = 0; i < size; i++) {
                Object item = nativeArray.get(i);
                String str = convertToString(item);
                if (str != null) {
                    list.add(str);
                }
            }
            return list;
        }

        if (data instanceof NativeObject) {
            // Single object - try to extract URL property
            String url = extractUrlFromObject((NativeObject) data);
            if (url != null) {
                List<String> list = new ArrayList<>(1);
                list.add(url);
                return list;
            }
            return new ArrayList<>();
        }

        if (data instanceof org.mozilla.javascript.NativeJavaObject) {
            // Unwrap the Java object
            org.mozilla.javascript.NativeJavaObject nativeJavaObj =
                    (org.mozilla.javascript.NativeJavaObject) data;
            Object unwrapped = nativeJavaObj.unwrap();
            if (unwrapped instanceof List) {
                @SuppressWarnings("unchecked")
                List<Object> objList = (List<Object>) unwrapped;
                List<String> list = new ArrayList<>(objList.size());
                for (Object item : objList) {
                    String str = convertToString(item);
                    if (str != null) {
                        list.add(str);
                    }
                }
                return list;
            }
            if (unwrapped != null && unwrapped.getClass().isArray()) {
                // Handle Java arrays
                Object[] array = (Object[]) unwrapped;
                List<String> list = new ArrayList<>(array.length);
                for (Object item : array) {
                    String str = convertToString(item);
                    if (str != null) {
                        list.add(str);
                    }
                }
                return list;
            }
            // Single object
            String url = convertToString(unwrapped);
            if (url != null) {
                List<String> list = new ArrayList<>(1);
                list.add(url);
                return list;
            }
            return new ArrayList<>();
        }

        if (data instanceof List) {
            @SuppressWarnings("unchecked")
            List<Object> objList = (List<Object>) data;
            List<String> list = new ArrayList<>(objList.size());
            for (Object item : objList) {
                String str = convertToString(item);
                if (str != null) {
                    list.add(str);
                }
            }
            return list;
        }

        // Single item - convert to string
        String str = convertToString(data);
        if (str != null) {
            List<String> list = new ArrayList<>(1);
            list.add(str);
            return list;
        }

        return new ArrayList<>();
    }

    /** Convert object to String URL. */
    private String convertToString(Object obj) {
        if (obj == null) {
            return null;
        }
        if (obj instanceof org.mozilla.javascript.Undefined) {
            return null;
        }
        String str = obj.toString();
        return (str != null && !str.isEmpty()) ? str : null;
    }

    /** Extract URL from NativeObject (checks common property names). */
    private String extractUrlFromObject(NativeObject obj) {
        String[] urlProps = {"url", "link", "href", "page", "pageUrl"};
        for (String prop : urlProps) {
            Object value = obj.get(prop, obj);
            // Check for NOT_FOUND (property doesn't exist)
            if (value == org.mozilla.javascript.UniqueTag.NOT_FOUND) {
                continue;
            }
            if (value != null && value != org.mozilla.javascript.Undefined.instance) {
                String url = convertToString(value);
                if (url != null && !url.isEmpty()) {
                    return url;
                }
            }
        }
        return null;
    }

    /** Extract base URL from full URL. */
    private String extractBaseUrl(String url) {
        if (url == null || url.isEmpty()) {
            return "";
        }
        try {
            java.net.URL parsedUrl = new java.net.URL(url);
            String protocol = parsedUrl.getProtocol();
            String host = parsedUrl.getHost();
            int port = parsedUrl.getPort();
            if (port == -1) {
                return protocol + "://" + host;
            }
            return protocol + "://" + host + ":" + port;
        } catch (Exception e) {
            return "";
        }
    }
}

File: engine/src/commonMain/kotlin/dev/haipham22/leechtext/plugin/js/loader/PageLoader.kt
package dev.haipham22.leechtext.plugin.js.loader

import dev.haipham22.leechtext.entities.PluginEntity
import dev.haipham22.leechtext.log.EngineLogger
import dev.haipham22.leechtext.plugin.js.api.JsValueBridge
import dev.haipham22.leechtext.plugin.js.sandbox.JsSandbox
import dev.haipham22.leechtext.plugin.js.sandbox.createSandbox
import dev.haipham22.leechtext.util.parseBaseUrl

/**
 * Loader page discovery (port từ loader/PageLoader.java; P5.2c: kết quả normalized,
 * NativeJavaObject unwrap đã làm ở boundary JsSandbox actual). Chạy script 'page',
 * trả về list URL các trang chapter — dùng cho site phân trang.
 */
class PageLoader private constructor(
    log: EngineLogger,
    plugin: PluginEntity,
) : AbstractLoader<List<String>>(log, plugin) {
    companion object {
        private val URL_PROPS = arrayOf("url", "link", "href", "page", "pageUrl")

        fun with(
            plugin: PluginEntity,
            log: EngineLogger,
        ): PageLoader = PageLoader(log, plugin)
    }

    override fun getScript(): String? = plugin.pageGetter

    override fun getLoaderType(): LoaderType = LoaderType.PAGE

    /**
     * One-shot discovery: trả về mọi page URL có sẵn. Empty list trên error (khác null của
     * AbstractLoader). Không set pluginSource (theo bản gốc — page script không cần load()).
     */
    override suspend fun load(url: String?): List<String> {
        if (getScript() == null) {
            log.add("[PageLoader] No script available")
            return ArrayList()
        }
        return super.load(url) ?: ArrayList()
    }

    @Suppress("ReturnCount") // guard-chain: từng bước fail thoát sớm, lifecycle sandbox/builder giữ nguyên shape
    override fun loadInSandbox(url: String?): List<String>? {
        val script = getScript() ?: return ArrayList()

        var sandbox: JsSandbox? = null
        try {
            // Extract base URL
            val baseUrl =
                plugin.source?.takeIf { it.isNotEmpty() } ?: parseBaseUrl(url ?: "")

            // Secure sandbox cho PAGE loader
            sandbox =
                createSandbox(log) {
                    loaderType = getLoaderType()
                    this.baseUrl = baseUrl
                    targetUrl = url
                }

            // Execute plugin script trong sandbox
            if (!sandbox.execute(script, getScriptName())) return ArrayList()

            // Call execute(url) (không có page parameter)
            val result = sandbox.callFunction("execute", url)
            if (result == null) return ArrayList()

            return processResult(result, url)
        } catch (e: Exception) {
            val errorMsg = e.message ?: e::class.simpleName
            log.add("[PageLoader] JavaScript execution error: $errorMsg")
            return ArrayList()
        } finally {
            sandbox?.close()
        }
    }

    override fun processResult(
        result: Any?,
        url: String?,
    ): List<String> {
        if (result == null) {
            log.add("[PageLoader] Script returned null or undefined")
            return ArrayList()
        }

        // Check Response wrapper (vBooks compatibility)
        if (!Response.isSuccess(result)) {
            val errorMsg = Response.getErrorMessage(result)
            log.add("[PageLoader] Error response: $errorMsg")
            return ArrayList()
        }

        val data = Response.getData(result)

        log.add("[PageLoader] Processing result, type: ${result::class.simpleName}")
        val urlList = convertToStringList(data)
        log.add("[PageLoader] Discovered ${urlList.size} page URLs")
        return urlList
    }

    /** Convert giá trị → List<String> URLs (List/Map single object/array-like). */
    private fun convertToStringList(data: Any?): List<String> {
        if (data == null) return ArrayList()

        if (data is List<*>) return itemsToStrings(data)

        if (data is Map<*, *>) {
            // Single object — thử extract property URL
            val url = extractUrlFromObject(data)
            return if (url != null) arrayListOf(url) else ArrayList()
        }

        if (JsValueBridge.sizeOf(data) >= 0) return itemsToStrings(JSResponse.convertArray(data))

        // Single item — convert sang string
        val str = convertToString(data)
        return if (str != null) arrayListOf(str) else ArrayList()
    }

    /** Iterable items → List<String> (bỏ item không stringify được). */
    private fun itemsToStrings(items: List<*>): List<String> {
        val list = ArrayList<String>(items.size)
        for (item in items) {
            convertToString(item)?.let { list.add(it) }
        }
        return list
    }

    /** Object → String URL. Giữ cả relative (contract PageLoaderTest) — chỉ loại rác
     * stringify của object gây "Expected URL scheme" xa hơn. */
    private fun convertToString(obj: Any?): String? {
        if (obj == null) return null
        if (obj is Map<*, *>) return null
        val str = JSResponse.getString(obj)?.trim() ?: return null
        if (str.isEmpty()) return null
        if (" " in str || str.contains("@") || str.contains("org.mozilla") || str.contains("com.dokar")) return null
        return str
    }

    /** Extract URL từ object (check các property name phổ biến). */
    private fun extractUrlFromObject(obj: Map<*, *>): String? {
        for (prop in URL_PROPS) {
            val value = obj[prop] ?: continue
            val url = convertToString(value)
            if (!url.isNullOrEmpty()) return url
        }
        return null
    }
}
